File: tools/synthetic_workout.py
#!/usr/bin/env python3
"""Build the reviewable FIT and visual fixture for the synthetic workout."""

from __future__ import annotations

import argparse
import math
import tempfile
from dataclasses import dataclass
from pathlib import Path

from fit_tool.base_type import BaseType
from fit_tool.developer_field import DeveloperField
from fit_tool.fit_file_builder import FitFileBuilder
from fit_tool.profile.messages.activity_message import ActivityMessage
from fit_tool.profile.messages.developer_data_id_message import DeveloperDataIdMessage
from fit_tool.profile.messages.field_description_message import FieldDescriptionMessage
from fit_tool.profile.messages.file_id_message import FileIdMessage
from fit_tool.profile.messages.lap_message import LapMessage
from fit_tool.profile.messages.record_message import RecordMessage
from fit_tool.profile.messages.session_message import SessionMessage
from fit_tool.profile.profile_type import FileType, Manufacturer


ROOT = Path(__file__).resolve().parents[1]
BASELINE = ROOT / "tools/baselines/synthetic-workout.svg"
START_MS = 1_700_000_000_000
SAMPLE_RATE = 25


@dataclass(frozen=True)
class Window:
    second: int
    phase: str
    style: str
    x: tuple[int, ...]
    y: tuple[int, ...]
    z: tuple[int, ...]
    rms: int
    peak: int
    crossings: int
    dynamic_rms: int
    dynamic_peak: int


def samples(style: str, second: int) -> tuple[tuple[int, ...], ...]:
    axes = [[], [], []]
    for index in range(SAMPLE_RATE):
        angle = 2 * math.pi * index / SAMPLE_RATE
        if style == "still":
            values = (0, 0, 1000)
        elif style == "smooth":
            amplitude = 560
            values = (
                round(amplitude * math.sin(angle)),
                round(amplitude * 0.55 * math.cos(angle)),
                round(1000 + amplitude * 0.22 * math.sin(2 * angle)),
            )
        elif style == "irregular":
            amplitude = 430 + ((second * 97 + index * 53) % 360)
            values = (
                round(amplitude * math.sin(angle + second * 0.17)),
                round(amplitude * 0.7 * math.cos(1.3 * angle)),
                round(1000 + amplitude * 0.3 * math.sin(2.4 * angle)),
            )
        elif style == "spike":
            values = (4200, -900, 1600) if index == 12 else (
                round(650 * math.sin(angle)),
                round(300 * math.cos(angle)),
                round(1000 + 160 * math.sin(2 * angle)),
            )
        else:
            raise ValueError(f"unknown motion style: {style}")
        for axis, value in zip(axes, values):
            axis.append(value)
    return tuple(tuple(axis) for axis in axes)


def features(x: tuple[int, ...], y: tuple[int, ...], z: tuple[int, ...]) -> tuple[int, ...]:
    magnitudes = [math.sqrt(a * a + b * b + c * c) for a, b, c in zip(x, y, z)]
    mean = sum(magnitudes) / len(magnitudes)
    means = tuple(sum(axis) / len(axis) for axis in (x, y, z))
    crossings = 0
    previous = 0
    for magnitude in magnitudes:
        sign = 1 if magnitude > mean else -1 if magnitude < mean else 0
        if sign and previous and sign != previous:
            crossings += 1
        if sign:
            previous = sign
    dynamic = [
        math.sqrt((a - means[0]) ** 2 + (b - means[1]) ** 2 + (c - means[2]) ** 2)
        for a, b, c in zip(x, y, z)
    ]
    return (
        int(math.sqrt(sum(value * value for value in magnitudes) / len(magnitudes))),
        int(max(magnitudes)),
        crossings,
        int(math.sqrt(sum(value * value for value in dynamic) / len(dynamic))),
        int(max(dynamic)),
    )


def workout() -> list[Window]:
    timeline = []
    for second in range(50):
        if second < 5 or 25 <= second < 30:
            phase, style = "rest", "still"
        elif second < 25:
            phase, style = "work", "smooth"
        else:
            phase, style = "work", "spike" if second == 42 else "irregular"
        x, y, z = samples(style, second)
        values = features(x, y, z)
        timeline.append(Window(second, phase, style, x, y, z, *values))
    return timeline


FIELD_TYPES = {
    0: ("total_sets", "sets", BaseType.UINT16, 2),
    1: ("battery_used", "%", BaseType.FLOAT32, 4),
    2: ("accel_rms", "mg", BaseType.UINT16, 2),
    3: ("accel_peak", "mg", BaseType.UINT16, 2),
    4: ("accel_zc", "crossings", BaseType.UINT8, 1),
    6: ("implement_count", "implements", BaseType.UINT8, 1),
    7: ("implement_weight", "g", BaseType.UINT16, 2),
    8: ("set_number", "set", BaseType.UINT16, 2),
    10: ("phase", "0=rest 1=work", BaseType.UINT8, 1),
    11: ("phase_duration", "s", BaseType.UINT16, 2),
    14: ("set_smoothness", "score", BaseType.SINT16, 2),
    17: ("total_swings", "swings", BaseType.UINT16, 2),
    18: ("swing_count", "swings", BaseType.UINT16, 2),
    19: ("motion_exposure", "mg-s", BaseType.UINT32, 4),
    20: ("motion_peak", "mg", BaseType.UINT16, 2),
    21: ("active_seconds", "s", BaseType.UINT16, 2),
    22: ("weight_volume", "kg-swings", BaseType.UINT32, 4),
    15: ("movement_type", "movement", BaseType.UINT8, 1),
    16: ("working_side", "side", BaseType.UINT8, 1),
    23: ("work_time", "s", BaseType.UINT32, 4),
    24: ("rest_time", "s", BaseType.UINT32, 4),
    26: ("implement_name", "", BaseType.STRING, 16),
}


def developer_field(field_id: int, value: int | float | str) -> DeveloperField:
    name, units, base_type, size = FIELD_TYPES[field_id]
    field = DeveloperField(
        developer_data_index=0, field_id=field_id, name=name,
        units=units, base_type=base_type, size=size,
    )
    field.set_value(0, value)
    return field


def build_fit(windows: list[Window], destination: Path) -> None:
    builder = FitFileBuilder(auto_define=True)
    file_id = FileIdMessage()
    file_id.type = FileType.ACTIVITY
    file_id.manufacturer = Manufacturer.DEVELOPMENT.value
    file_id.product = 1
    file_id.serial_number = 424242
    file_id.time_created = START_MS
    builder.add(file_id)

    developer = DeveloperDataIdMessage()
    developer.developer_data_index = 0
    developer.application_id = bytes.fromhex("6f0f19e1a0e14842a7b70ac011223344")
    developer.application_version = 999
    builder.add(developer)
    for field_id, (name, units, base_type, _) in FIELD_TYPES.items():
        description = FieldDescriptionMessage()
        description.developer_data_index = 0
        description.field_definition_number = field_id
        description.fit_base_type_id = base_type
        description.field_name = name
        description.units = units
        builder.add(description)

    for window in windows:
        record = RecordMessage(developer_fields=[
            developer_field(2, window.rms),
            developer_field(3, window.peak),
            developer_field(4, window.crossings),
        ])
        record.timestamp = START_MS + window.second * 1000
        record.heart_rate = 68 + round(window.second * 0.65) + (7 if window.phase == "work" else 0)
        builder.add(record)

    laps = [
        (0, 5, 0, 0, -1, 0),
        (5, 20, 1, 1, 88, 20),
        (25, 5, 0, 0, -1, 0),
        (30, 20, 1, 2, 61, 17),
    ]
    for index, (start, duration, phase, set_number, smoothness, swings) in enumerate(laps):
        segment = windows[start:start + duration]
        lap = LapMessage(developer_fields=[
            developer_field(8, set_number), developer_field(10, phase),
            developer_field(11, duration), developer_field(14, smoothness),
            developer_field(15, 4), developer_field(16, 3),
            developer_field(18, swings),
            developer_field(19, sum(item.dynamic_rms for item in segment)),
            developer_field(20, max(item.dynamic_peak for item in segment)),
            developer_field(21, duration if phase else 0),
            developer_field(22, 8 * swings if phase else 0),
        ])
        lap.message_index = index
        lap.start_time = START_MS + start * 1000
        lap.timestamp = START_MS + (start + duration) * 1000
        lap.total_elapsed_time = duration
        lap.total_timer_time = duration
        builder.add(lap)

    session = SessionMessage(developer_fields=[
        developer_field(0, 2), developer_field(1, 1.5),
        developer_field(6, 2), developer_field(7, 4000),
        developer_field(17, 37), developer_field(23, 40), developer_field(24, 10),
        developer_field(26, "Clubs: 2 x 4 kg"),
    ])
    session.start_time = START_MS
    session.timestamp = START_MS + 50_000
    session.total_elapsed_time = 50
    session.total_timer_time = 50
    session.avg_heart_rate = 91
    session.max_heart_rate = 107
    session.num_laps = len(laps)
    builder.add(session)

    activity = ActivityMessage()
    activity.timestamp = START_MS + 50_000
    activity.total_timer_time = 50
    activity.num_sessions = 1
    builder.add(activity)
    destination.parent.mkdir(parents=True, exist_ok=True)
    builder.build().to_file(str(destination))


def render_svg(windows: list[Window]) -> str:
    width, height = 960, 580
    left, right, top, chart_bottom = 72, 28, 92, 400
    chart_width = width - left - right
    maximum = max(item.peak for item in windows)
    x = lambda second: left + chart_width * second / (len(windows) - 1)
    y = lambda value: chart_bottom - (chart_bottom - top) * value / maximum
    points = lambda key: " ".join(f"{x(item.second):.1f},{y(getattr(item, key)):.1f}" for item in windows)
    backgrounds = []
    for start, duration, phase in ((0, 5, "Rest"), (5, 20, "Smooth"), (25, 5, "Rest"), (30, 20, "Irregular")):
        color = "#e9f0f7" if phase == "Rest" else "#fff0e9"
        backgrounds.append(
            f'<rect x="{x(start):.1f}" y="{top}" width="{chart_width * duration / 50:.1f}" '
            f'height="{chart_bottom-top}" fill="{color}"/><text x="{x(start)+8:.1f}" y="{top+20}" '
            f'font-size="13" fill="#6f6961">{phase}</text>'
        )
    bars = []
    for set_number, score, start in ((1, 88, 5), (2, 61, 30)):
        bar_x = 220 + (set_number - 1) * 260
        bar_height = score * 0.65
        bars.append(
            f'<rect x="{bar_x}" y="{565-bar_height:.1f}" width="110" height="{bar_height:.1f}" '
            f'rx="3" fill="#397a68"/><text x="{bar_x+55}" y="{553-bar_height:.1f}" '
            f'text-anchor="middle" font-size="16" fill="#24211d">{score}</text><text x="{bar_x+55}" '
            f'y="578" text-anchor="middle" font-size="13" fill="#6f6961">Set {set_number}</text>'
        )
    return f'''<svg xmlns="http://www.w3.org/2000/svg" width="{width}" height="{height}" viewBox="0 0 {width} {height}">
<rect width="100%" height="100%" fill="#f7f5f1"/><text x="{left}" y="38" font-family="system-ui,sans-serif" font-size="25" font-weight="600" fill="#24211d">Synthetic motion confidence check</text>
<text x="{left}" y="64" font-family="system-ui,sans-serif" font-size="14" fill="#6f6961">Stillness → smooth periodic swings → rest → irregular swings with one deliberate spike</text>
<g font-family="system-ui,sans-serif">{''.join(backgrounds)}
<line x1="{left}" y1="{chart_bottom}" x2="{width-right}" y2="{chart_bottom}" stroke="#bdb7af"/>
<polyline points="{points('rms')}" fill="none" stroke="#bd3e14" stroke-width="3"/>
<polyline points="{points('peak')}" fill="none" stroke="#e4a02b" stroke-width="2"/>
<text x="{left}" y="425" font-size="13" fill="#bd3e14">Acceleration RMS</text><text x="220" y="425" font-size="13" fill="#b37813">Peak acceleration</text>
<text x="{left}" y="466" font-size="18" font-weight="600" fill="#24211d">Expected smoothness contrast</text>{''.join(bars)}</g></svg>'''


def generated_outputs(directory: Path) -> dict[Path, str | bytes]:
    windows = workout()
    fit_path = directory / "synthetic-workout.fit"
    build_fit(windows, fit_path)
    return {
        fit_path: fit_path.read_bytes(),
        directory / "synthetic-workout.svg": render_svg(windows),
    }


def main(argv=None) -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--output-dir", type=Path, default=Path("build/synthetic-workout"))
    parser.add_argument("--check", action="store_true")
    args = parser.parse_args(argv)
    with tempfile.TemporaryDirectory(prefix="mace-synthetic-") as temporary:
        outputs = generated_outputs(Path(temporary))
        svg = next(value for path, value in outputs.items() if path.suffix == ".svg")
        if args.check:
            failures = []
            if not BASELINE.exists() or BASELINE.read_text() != svg:
                failures.append(f"visual baseline differs: regenerate {BASELINE}")
            if failures:
                raise SystemExit("\n".join(failures))
        args.output_dir.mkdir(parents=True, exist_ok=True)
        for source, content in outputs.items():
            destination = args.output_dir / source.name
            destination.write_bytes(content) if isinstance(content, bytes) else destination.write_text(content)
    print(args.output_dir)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/synthetic-workout.ts
#!/usr/bin/env node
// Build the reviewable FIT and visual fixture for the synthetic workout.

import { existsSync, mkdirSync, mkdtempSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { basename, dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import {
    type ActivityMesg,
    type DeveloperDataIdMesg,
    type Encodable,
    Encoder,
    type FieldDescription,
    type FieldDescriptionMesg,
    type FileIdMesg,
    type LapMesg,
    Profile,
    type RecordMesg,
    type SessionMesg,
} from "@garmin/fitsdk";

import { pythonRound } from "./fit-io.ts";

export const BASELINE = fileURLToPath(new URL("baselines/synthetic-workout.svg", import.meta.url));
export const START_MS = 1_700_000_000_000;
const SAMPLE_RATE = 25;

export type Phase = "rest" | "work";
export type Style = "still" | "smooth" | "irregular" | "spike";

export interface Window {
    second: number;
    phase: Phase;
    style: Style;
    x: number[];
    y: number[];
    z: number[];
    rms: number;
    peak: number;
    crossings: number;
    dynamic_rms: number;
    dynamic_peak: number;
}

export function samples(style: string, second: number): [number[], number[], number[]] {
    const axes: [number[], number[], number[]] = [[], [], []];
    for (let index = 0; index < SAMPLE_RATE; index++) {
        const angle = (2 * Math.PI * index) / SAMPLE_RATE;
        let values: [number, number, number];
        if (style === "still") {
            values = [0, 0, 1000];
        } else if (style === "smooth") {
            const amplitude = 560;
            values = [
                pythonRound(amplitude * Math.sin(angle)),
                pythonRound(amplitude * 0.55 * Math.cos(angle)),
                pythonRound(1000 + amplitude * 0.22 * Math.sin(2 * angle)),
            ];
        } else if (style === "irregular") {
            const amplitude = 430 + ((second * 97 + index * 53) % 360);
            values = [
                pythonRound(amplitude * Math.sin(angle + second * 0.17)),
                pythonRound(amplitude * 0.7 * Math.cos(1.3 * angle)),
                pythonRound(1000 + amplitude * 0.3 * Math.sin(2.4 * angle)),
            ];
        } else if (style === "spike") {
            values = index === 12 ? [4200, -900, 1600] : [
                pythonRound(650 * Math.sin(angle)),
                pythonRound(300 * Math.cos(angle)),
                pythonRound(1000 + 160 * Math.sin(2 * angle)),
            ];
        } else {
            throw new Error(`unknown motion style: ${style}`);
        }
        axes[0].push(values[0]);
        axes[1].push(values[1]);
        axes[2].push(values[2]);
    }
    return axes;
}

const square = (value: number): number => value * value;

function features(x: number[], y: number[], z: number[]): [number, number, number, number, number] {
    const magnitudes = x.map((value, index) =>
        Math.sqrt(value * value + (y[index] ?? 0) * (y[index] ?? 0) + (z[index] ?? 0) * (z[index] ?? 0)));
    const mean = magnitudes.reduce((total, value) => total + value, 0) / magnitudes.length;
    const means = [x, y, z].map((axis) => axis.reduce((total, value) => total + value, 0) / axis.length) as
        [number, number, number];
    let crossings = 0;
    let previous = 0;
    for (const magnitude of magnitudes) {
        const sign = magnitude > mean ? 1 : magnitude < mean ? -1 : 0;
        if (sign !== 0 && previous !== 0 && sign !== previous) {
            crossings += 1;
        }
        if (sign !== 0) {
            previous = sign;
        }
    }
    const dynamic = x.map((value, index) =>
        Math.sqrt(square(value - means[0]) + square((y[index] ?? 0) - means[1]) + square((z[index] ?? 0) - means[2])));
    const rootMeanSquare = (values: number[]): number =>
        Math.trunc(Math.sqrt(values.reduce((total, value) => total + value * value, 0) / values.length));
    return [
        rootMeanSquare(magnitudes),
        Math.trunc(Math.max(...magnitudes)),
        crossings,
        rootMeanSquare(dynamic),
        Math.trunc(Math.max(...dynamic)),
    ];
}

export function workout(): Window[] {
    const timeline: Window[] = [];
    for (let second = 0; second < 50; second++) {
        let phase: Phase;
        let style: Style;
        if (second < 5 || (second >= 25 && second < 30)) {
            [phase, style] = ["rest", "still"];
        } else if (second < 25) {
            [phase, style] = ["work", "smooth"];
        } else {
            [phase, style] = ["work", second === 42 ? "spike" : "irregular"];
        }
        const [x, y, z] = samples(style, second);
        const [rms, peak, crossings, dynamicRms, dynamicPeak] = features(x, y, z);
        timeline.push({
            second, phase, style, x, y, z,
            rms, peak, crossings,
            dynamic_rms: dynamicRms, dynamic_peak: dynamicPeak,
        });
    }
    return timeline;
}

// Profile.MesgNum is typed as an index signature; resolve the handful of
// message numbers this fixture writes once, with a hard failure on typos.
function mesgNum(name: string): number {
    const value = Profile.MesgNum[name];
    if (value === undefined) {
        throw new Error(`unknown FIT message type: ${name}`);
    }
    return value;
}

const MESG = {
    fileId: mesgNum("FILE_ID"),
    developerDataId: mesgNum("DEVELOPER_DATA_ID"),
    fieldDescription: mesgNum("FIELD_DESCRIPTION"),
    record: mesgNum("RECORD"),
    lap: mesgNum("LAP"),
    session: mesgNum("SESSION"),
    activity: mesgNum("ACTIVITY"),
} as const;

// FIT wire values for fit_base_type_id (profile type fitBaseType); the SDK's
// Utils.FitBaseType indexes are NOT valid field_description content.
const BASE_TYPE = { uint8: 2, string: 7, sint16: 131, uint16: 132, uint32: 134, float32: 136 } as const;

const FIELD_TYPES: readonly [number, string, string, number][] = [
    [0, "total_sets", "sets", BASE_TYPE.uint16],
    [1, "battery_used", "%", BASE_TYPE.float32],
    [2, "accel_rms", "mg", BASE_TYPE.uint16],
    [3, "accel_peak", "mg", BASE_TYPE.uint16],
    [4, "accel_zc", "crossings", BASE_TYPE.uint8],
    [6, "implement_count", "implements", BASE_TYPE.uint8],
    [7, "implement_weight", "g", BASE_TYPE.uint16],
    [8, "set_number", "set", BASE_TYPE.uint16],
    [10, "phase", "0=rest 1=work", BASE_TYPE.uint8],
    [11, "phase_duration", "s", BASE_TYPE.uint16],
    [14, "set_smoothness", "score", BASE_TYPE.sint16],
    [17, "total_swings", "swings", BASE_TYPE.uint16],
    [18, "swing_count", "swings", BASE_TYPE.uint16],
    [19, "motion_exposure", "mg-s", BASE_TYPE.uint32],
    [20, "motion_peak", "mg", BASE_TYPE.uint16],
    [21, "active_seconds", "s", BASE_TYPE.uint16],
    [22, "weight_volume", "kg-swings", BASE_TYPE.uint32],
    [15, "movement_type", "movement", BASE_TYPE.uint8],
    [16, "working_side", "side", BASE_TYPE.uint8],
    [23, "work_time", "s", BASE_TYPE.uint32],
    [24, "rest_time", "s", BASE_TYPE.uint32],
    [26, "implement_name", "", BASE_TYPE.string],
    [27, "swing_total", "swings", BASE_TYPE.uint16],
    [28, "swing_event", "swings", BASE_TYPE.uint8],
    [29, "swing_cadence", "spm", BASE_TYPE.uint8],
    [30, "smoothness_score", "score", BASE_TYPE.uint8],
];

const DEVELOPER_DATA_ID_MESG: Encodable<DeveloperDataIdMesg> = {
    mesgNum: MESG.developerDataId,
    applicationId: Array.from(Buffer.from("6f0f19e1a0e14842a7b70ac011223344", "hex")),
    applicationVersion: 999,
    developerDataIndex: 0,
};

const FIELD_DESCRIPTION_MESGS: readonly Encodable<FieldDescriptionMesg>[] = FIELD_TYPES.map(
    ([fieldId, name, units, baseType]) => ({
        mesgNum: MESG.fieldDescription,
        developerDataIndex: 0,
        fieldDefinitionNumber: fieldId,
        fitBaseTypeId: baseType,
        fieldName: name,
        units,
    }),
);

// Developer fields are keyed by field definition number for both the encoder
// registration and per-message values.
const FIELD_DESCRIPTIONS: Record<number, FieldDescription> = Object.fromEntries(
    FIELD_DESCRIPTION_MESGS.map((mesg) => [
        mesg.fieldDefinitionNumber ?? 0,
        { developerDataIdMesg: DEVELOPER_DATA_ID_MESG, fieldDescriptionMesg: mesg },
    ]),
);

export function buildFit(windows: readonly Window[], destination: string): void {
    const encoder = new Encoder({ fieldDescriptions: FIELD_DESCRIPTIONS });
    const fileId: Encodable<FileIdMesg> = {
        mesgNum: MESG.fileId,
        type: "activity",
        manufacturer: "development",
        product: 1,
        serialNumber: 424242,
        timeCreated: new Date(START_MS),
    };
    encoder.writeMesg(fileId);
    encoder.writeMesg(DEVELOPER_DATA_ID_MESG);
    for (const mesg of FIELD_DESCRIPTION_MESGS) {
        encoder.writeMesg(mesg);
    }

    let swingTotal = 0;
    let recentEvents: number[] = [];
    let smoothReference: [number, number, number] | null = null;
    let smoothTotal = 0;
    let smoothWindows = 0;
    let validWindows = 0;
    for (const window of windows) {
        // The committed fixture mirrors the production record contract. Its
        // deterministic event shape is presentation data; Monkey C tests own
        // the detector and production-pipeline assertions.
        const event = window.phase === "work" && window.second % 2 === 0 ? 1 : 0;
        swingTotal += event;
        recentEvents.push(event);
        recentEvents = recentEvents.slice(-10);
        const cadence = pythonRound(
            (recentEvents.reduce((total, value) => total + value, 0) * 60) / recentEvents.length);
        if (window.phase === "work" && window.dynamic_rms >= 40) {
            if (smoothReference === null) {
                smoothReference = [window.dynamic_rms, window.dynamic_peak, window.crossings];
            } else {
                if (validWindows >= 4) {
                    const differences = [
                        Math.abs(window.dynamic_rms - smoothReference[0]) / Math.max(Math.abs(smoothReference[0]), 40),
                        Math.abs(window.dynamic_peak - smoothReference[1]) / Math.max(Math.abs(smoothReference[1]), 80),
                        Math.abs(window.crossings - smoothReference[2]) / Math.max(Math.abs(smoothReference[2]), 2),
                    ] as [number, number, number];
                    smoothTotal += Math.max(0, Math.min(100, Math.trunc(
                        100 - 100 * (0.45 * differences[0] + 0.35 * differences[1] + 0.20 * differences[2]))));
                    smoothWindows += 1;
                }
                smoothReference = [
                    0.8 * smoothReference[0] + 0.2 * window.dynamic_rms,
                    0.8 * smoothReference[1] + 0.2 * window.dynamic_peak,
                    0.8 * smoothReference[2] + 0.2 * window.crossings,
                ];
            }
            validWindows += 1;
        }
        const smoothScore = smoothWindows !== 0 ? pythonRound(smoothTotal / smoothWindows) : 0;
        const record: Encodable<RecordMesg> = {
            mesgNum: MESG.record,
            timestamp: new Date(START_MS + window.second * 1000),
            heartRate: 68 + pythonRound(window.second * 0.65) + (window.phase === "work" ? 7 : 0),
            developerFields: {
                2: window.rms,
                3: window.peak,
                4: window.crossings,
                27: swingTotal,
                28: event,
                29: cadence,
                30: smoothScore,
            },
        };
        encoder.writeMesg(record);
    }

    const laps: readonly [number, number, number, number, number, number][] = [
        [0, 5, 0, 0, -1, 0],
        [5, 20, 1, 1, 88, 20],
        [25, 5, 0, 0, -1, 0],
        [30, 20, 1, 2, 61, 17],
    ];
    laps.forEach(([start, duration, phase, setNumber, smoothness, swings], index) => {
        const segment = windows.slice(start, start + duration);
        const lap: Encodable<LapMesg> = {
            mesgNum: MESG.lap,
            messageIndex: index,
            startTime: new Date(START_MS + start * 1000),
            timestamp: new Date(START_MS + (start + duration) * 1000),
            totalElapsedTime: duration,
            totalTimerTime: duration,
            developerFields: {
                8: setNumber,
                10: phase,
                11: duration,
                14: smoothness,
                15: 4,
                16: 3,
                18: swings,
                19: segment.reduce((total, item) => total + item.dynamic_rms, 0),
                20: Math.max(...segment.map((item) => item.dynamic_peak)),
                21: phase !== 0 ? duration : 0,
                22: phase !== 0 ? 8 * swings : 0,
            },
        };
        encoder.writeMesg(lap);
    });

    const session: Encodable<SessionMesg> = {
        mesgNum: MESG.session,
        startTime: new Date(START_MS),
        timestamp: new Date(START_MS + 50_000),
        totalElapsedTime: 50,
        totalTimerTime: 50,
        avgHeartRate: 91,
        maxHeartRate: 107,
        numLaps: laps.length,
        developerFields: {
            0: 2,
            1: 1.5,
            6: 2,
            7: 4000,
            17: 37,
            23: 40,
            24: 10,
            26: "Clubs: 2 x 4 kg",
        },
    };
    encoder.writeMesg(session);
    const activity: Encodable<ActivityMesg> = {
        mesgNum: MESG.activity,
        timestamp: new Date(START_MS + 50_000),
        totalTimerTime: 50,
        numSessions: 1,
    };
    encoder.writeMesg(activity);

    mkdirSync(dirname(destination), { recursive: true });
    writeFileSync(destination, encoder.close());
}

const fixed1 = (value: number): string => value.toFixed(1);

export function renderSvg(windows: readonly Window[]): string {
    const [width, height] = [960, 580];
    const [left, right, top, chartBottom] = [72, 28, 92, 400];
    const chartWidth = width - left - right;
    const maximum = Math.max(...windows.map((item) => item.peak));
    const x = (second: number): number => left + (chartWidth * second) / (windows.length - 1);
    const y = (value: number): number => chartBottom - ((chartBottom - top) * value) / maximum;
    const points = (key: "rms" | "peak"): string =>
        windows.map((item) => `${fixed1(x(item.second))},${fixed1(y(item[key]))}`).join(" ");
    const backgrounds: string[] = [];
    const phases: readonly [number, number, string][] = [
        [0, 5, "Rest"], [5, 20, "Smooth"], [25, 5, "Rest"], [30, 20, "Irregular"],
    ];
    for (const [start, duration, phase] of phases) {
        const color = phase === "Rest" ? "#e9f0f7" : "#fff0e9";
        backgrounds.push(
            `<rect x="${fixed1(x(start))}" y="${String(top)}" width="${fixed1((chartWidth * duration) / 50)}" `
            + `height="${String(chartBottom - top)}" fill="${color}"/><text x="${fixed1(x(start) + 8)}" `
            + `y="${String(top + 20)}" `
            + `font-size="13" fill="#6f6961">${phase}</text>`,
        );
    }
    const bars: string[] = [];
    const scores: readonly [number, number][] = [[1, 88], [2, 61]];
    for (const [setNumber, score] of scores) {
        const barX = 220 + (setNumber - 1) * 260;
        const barHeight = score * 0.65;
        bars.push(
            `<rect x="${String(barX)}" y="${fixed1(565 - barHeight)}" width="110" height="${fixed1(barHeight)}" `
            + `rx="3" fill="#397a68"/><text x="${String(barX + 55)}" y="${fixed1(553 - barHeight)}" `
            + `text-anchor="middle" font-size="16" fill="#24211d">${String(score)}</text><text x="${String(barX + 55)}" `
            + `y="578" text-anchor="middle" font-size="13" fill="#6f6961">Set ${String(setNumber)}</text>`,
        );
    }
    return `<svg xmlns="http://www.w3.org/2000/svg" width="${String(width)}" height="${String(height)}" viewBox="0 0 ${String(width)} ${String(height)}">
<rect width="100%" height="100%" fill="#f7f5f1"/><text x="${String(left)}" y="38" font-family="system-ui,sans-serif" font-size="25" font-weight="600" fill="#24211d">Synthetic motion confidence check</text>
<text x="${String(left)}" y="64" font-family="system-ui,sans-serif" font-size="14" fill="#6f6961">Stillness → smooth periodic swings → rest → irregular swings with one deliberate spike</text>
<g font-family="system-ui,sans-serif">${backgrounds.join("")}
<line x1="${String(left)}" y1="${String(chartBottom)}" x2="${String(width - right)}" y2="${String(chartBottom)}" stroke="#bdb7af"/>
<polyline points="${points("rms")}" fill="none" stroke="#bd3e14" stroke-width="3"/>
<polyline points="${points("peak")}" fill="none" stroke="#e4a02b" stroke-width="2"/>
<text x="${String(left)}" y="425" font-size="13" fill="#bd3e14">Acceleration RMS</text><text x="220" y="425" font-size="13" fill="#b37813">Peak acceleration</text>
<text x="${String(left)}" y="466" font-size="18" font-weight="600" fill="#24211d">Expected smoothness contrast</text>${bars.join("")}</g></svg>`;
}

export function generatedOutputs(directory: string): Map<string, Buffer | string> {
    const windows = workout();
    const fitDestination = join(directory, "synthetic-workout.fit");
    buildFit(windows, fitDestination);
    return new Map<string, Buffer | string>([
        [fitDestination, readFileSync(fitDestination)],
        [join(directory, "synthetic-workout.svg"), renderSvg(windows)],
    ]);
}

interface CliArgs {
    outputDir: string;
    check: boolean;
}

function parseArgs(argv: readonly string[]): CliArgs {
    let outputDir = join("build", "synthetic-workout");
    let check = false;
    for (let index = 0; index < argv.length; index++) {
        const argument = argv[index];
        if (argument === "--output-dir") {
            index++;
            outputDir = argv[index] ?? outputDir;
        } else if (argument === "--check") {
            check = true;
        } else {
            throw new Error(`unexpected argument: ${String(argument)}`);
        }
    }
    return { outputDir, check };
}

export function main(argv: readonly string[] = process.argv.slice(2), baseline: string = BASELINE): number {
    const args = parseArgs(argv);
    const temporary = mkdtempSync(join(tmpdir(), "mace-synthetic-"));
    try {
        const outputs = generatedOutputs(temporary);
        const svg = [...outputs.entries()].find(([path]) => path.endsWith(".svg"))?.[1];
        if (args.check && (!existsSync(baseline) || readFileSync(baseline, "utf8") !== svg)) {
            throw new Error(`visual baseline differs: regenerate ${baseline}`);
        }
        mkdirSync(args.outputDir, { recursive: true });
        for (const [source, content] of outputs) {
            writeFileSync(join(args.outputDir, basename(source)), content);
        }
    } finally {
        rmSync(temporary, { recursive: true, force: true });
    }
    console.log(args.outputDir);
    return 0;
}

if (process.argv[1] === fileURLToPath(import.meta.url)) {
    try {
        process.exitCode = main();
    } catch (error) {
        console.error(error instanceof Error ? error.message : String(error));
        process.exitCode = 1;
    }
}
